File: frontend/src/lib/conventions/idempotency.ts
// Hand-written Idempotency-Key helper, dual to backend/internal/shared/idx/
// idempotency.go. Wire format: `v1.{unixSeconds}.{uuidv7}`. Spec: 00-shared-
// conventions §3.4 (24h TTL); shared-conventions-codified spec §6 C-4.

import { newId, requireServerId, UUID_V7_REGEX } from '../ids';
import { IDEMPOTENCY_KEY_TTL_SECONDS } from './pagination';

/** Wire-format version. Bumping signals an incompatible format change. */
export const IDEMPOTENCY_KEY_VERSION = 'v1';

const DECIMAL_UNIX_SECONDS = /^\d+$/;

export interface IdempotencyKey {
  version: string;
  issuedAt: Date;
  uuid: string;
}

/** Returns a fresh Idempotency-Key valid for {@link IDEMPOTENCY_KEY_TTL_SECONDS} seconds. */
export function generateIdempotencyKey(now: Date = new Date()): string {
  return formatIdempotencyKey(now, newId());
}

/** Formats an arbitrary timestamp + UUIDv7 into the wire format. */
export function formatIdempotencyKey(issuedAt: Date, uuid: string): string {
  const unixSeconds = Math.floor(issuedAt.getTime() / 1000);
  return `${IDEMPOTENCY_KEY_VERSION}.${unixSeconds}.${uuid}`;
}

/**
 * Validates and decodes a key without checking expiry. Throws when the input
 * is empty, the version is wrong, the timestamp is non-numeric, or the UUID
 * portion is not a valid UUIDv7 (including any value with the `tmp_` prefix).
 */
export function parseIdempotencyKey(raw: string): IdempotencyKey {
  if (!raw) {
    throw new Error('parseIdempotencyKey: empty');
  }
  const parts = raw.split('.');
  if (parts.length !== 3) {
    throw new Error(`parseIdempotencyKey: expected 3 parts, got ${parts.length}`);
  }
  // Length check above narrows the array contract; cast lets noUncheckedIndexedAccess pass.
  const [version, unixStr, uuid] = parts as [string, string, string];
  if (version !== IDEMPOTENCY_KEY_VERSION) {
    throw new Error(`parseIdempotencyKey: version ${version} (only ${IDEMPOTENCY_KEY_VERSION} accepted)`);
  }
  if (!DECIMAL_UNIX_SECONDS.test(unixStr)) {
    throw new Error(`parseIdempotencyKey: timestamp not decimal Unix seconds: ${unixStr}`);
  }
  const unixSec = Number(unixStr);
  if (!Number.isSafeInteger(unixSec)) {
    throw new Error(`parseIdempotencyKey: timestamp outside safe integer range: ${unixStr}`);
  }
  // requireServerId throws on tmp_ prefix and on non-UUIDv7 strings, dual to Go side.
  requireServerId(uuid);
  if (!UUID_V7_REGEX.test(uuid)) {
    // Defensive: requireServerId already enforces this, but keep the explicit check.
    throw new Error(`parseIdempotencyKey: bad uuid ${uuid}`);
  }
  return {
    version,
    issuedAt: new Date(unixSec * 1000),
    uuid,
  };
}

/** True when the key was issued more than {@link IDEMPOTENCY_KEY_TTL_SECONDS} ago. */
export function isIdempotencyKeyExpired(raw: string, now: Date = new Date()): boolean {
  const parsed = parseIdempotencyKey(raw);
  const ageSeconds = (now.getTime() - parsed.issuedAt.getTime()) / 1000;
  return ageSeconds > IDEMPOTENCY_KEY_TTL_SECONDS;
}

File: backend/internal/shared/idx/idempotency.go
package idx

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// IdempotencyKeyVersion is the wire-format version prefix for keys produced by
// GenerateIdempotencyKey. Bumping it indicates a backward-incompatible change
// to either the format or the semantics of expiry; the parser only accepts
// keys that match the current version.
const IdempotencyKeyVersion = "v1"

// ErrIdempotencyKeyMalformed is returned when the input is not a syntactically
// valid Idempotency-Key string for the current version.
var ErrIdempotencyKeyMalformed = errors.New("idx: idempotency key malformed")

// IdempotencyKey is the parsed representation of a wire-format key. The
// embedded UUID is the request-scoped UUIDv7 originally minted on the client
// or server; IssuedAt is the explicit timestamp prefix, used for TTL checks
// without parsing the UUIDv7 timestamp bytes.
type IdempotencyKey struct {
	Version  string
	IssuedAt time.Time
	UUID     string
}

// GenerateIdempotencyKey returns a freshly minted key valid until
// IssuedAt + IdempotencyKeyTTLSeconds. The wire format is
// `v1.{unixSeconds}.{uuidv7}` so the TS side can parse it without any
// UUIDv7 byte unpacking.
func GenerateIdempotencyKey() string {
	return formatIdempotencyKey(time.Now(), NewID())
}

func formatIdempotencyKey(issuedAt time.Time, uuid string) string {
	return fmt.Sprintf("%s.%d.%s", IdempotencyKeyVersion, issuedAt.Unix(), uuid)
}

// ParseIdempotencyKey validates and decodes a key without checking expiry.
// It rejects empty input, wrong version, non-numeric timestamps, non-UUIDv7
// bodies, and any value carrying the browser-only `tmp_` prefix.
func ParseIdempotencyKey(key string) (IdempotencyKey, error) {
	if key == "" {
		return IdempotencyKey{}, fmt.Errorf("%w: empty", ErrIdempotencyKeyMalformed)
	}
	parts := strings.Split(key, ".")
	if len(parts) != 3 {
		return IdempotencyKey{}, fmt.Errorf("%w: expected 3 dot-separated parts, got %d", ErrIdempotencyKeyMalformed, len(parts))
	}
	if parts[0] != IdempotencyKeyVersion {
		return IdempotencyKey{}, fmt.Errorf("%w: version %q (only %q is accepted)", ErrIdempotencyKeyMalformed, parts[0], IdempotencyKeyVersion)
	}
	unixSec, err := strconv.ParseInt(parts[1], 10, 64)
	if err != nil {
		return IdempotencyKey{}, fmt.Errorf("%w: timestamp not numeric: %s", ErrIdempotencyKeyMalformed, parts[1])
	}
	if err := RequireServerID(parts[2]); err != nil {
		return IdempotencyKey{}, fmt.Errorf("%w: %v", ErrIdempotencyKeyMalformed, err)
	}
	return IdempotencyKey{
		Version:  parts[0],
		IssuedAt: time.Unix(unixSec, 0),
		UUID:     parts[2],
	}, nil
}

// IsIdempotencyKeyExpired returns true when the key was issued more than
// IdempotencyKeyTTLSeconds ago relative to `now`. Pass time.Now() in
// production; tests inject a fixed clock.
func IsIdempotencyKeyExpired(key string, now time.Time) (bool, error) {
	parsed, err := ParseIdempotencyKey(key)
	if err != nil {
		return false, err
	}
	age := now.Sub(parsed.IssuedAt)
	return age > time.Duration(IdempotencyKeyTTLSeconds)*time.Second, nil
}
